Add before and since filter to images

vdemeester opened the pull request:
Add support for two now filter on the `images` command : `before` and `since`. They work the same as the one on the `ps` command but for images 🐍.

```
    $ docker images --filter before=myimage
    # display all images older than myimage
    $ docker images --filter since=myimage
    # display all images younger than myimage
```

Closes #7967 :angel: 

**\- Description for the changelog**

Add `before` and `since` filters for `docker images` command.

🐸

Signed-off-by: Vincent Demeester vincent@sbr.pm

cpuguy83 commented:
Design OK for me.

cpuguy83 commented:
Sorry, thought this was using `--filter` at first, I'd rather keep the arguments in there instead of creating new args for them....

vdemeester commented:
@cpuguy83 this is using `--filter`.. I'm just that tired I messed up in the description :scream:

thaJeztah commented:
Design LGTM (but the description here at the top is incorrect; it's `--filter before myimage`, not `--before`

thaJeztah commented:
oh! GitHub is slow at the moment, missed your comment :D

cpuguy83 commented on integration-cli/docker_cli_images_test.go:
```diff
@@ -121,6 +121,76 @@ func (s *DockerSuite) TestImagesFilterLabelWithCommit(c *check.C) {
 }
 
+func (s *DockerSuite) TestImagesFilterSinceAndBefore(c *check.C) {
+	testRequires(c, DaemonIsLinux)
```
why `DaemonIsLinux`?

vdemeester replied:
Because of the build from scratch which is not available for windoS am I right ? :angel:

cpuguy83 replied:
I think there is a var for this like "minimalBaseImage" or something.

thaJeztah replied:
yes https://github.com/docker/docker/blob/3d13fddd2bc4d679f0eaa68b0be877e5a816ad53/integration-cli/docker_cli_history_test.go#L21

vdemeester replied:
ah coll, I'll update then :wink:

vdemeester replied:
updated :angel:

cpuguy83 commented:
LGTM

thaJeztah commented:
LGTM, tested it and works as advertised. Moving to docs review

thaJeztah commented on man/docker-images.1.md:
```diff
@@ -38,7 +38,9 @@ versions.
+   Filters the output based on these conditions:
+   - dangling=(true|false) - finds unused images.
+   - label=<key> or label=<key>=<value>
 
```
You forgot to add the new options here

vdemeester replied:
:sweat_smile:

thaJeztah commented:
Looks like you have to update the API docs as well;

- https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api_v1.24.md#list-images
- and https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api.md#v124-api-changes

vdemeester commented:
Also ping @albers @sdurrheimer for bash/zsh completion review :angel:

albers commented:
I'm confused by the term "image reference", haven't heard this before.

vdemeester commented:
@albers hum you are right, mentioning here and not elsewhere might through people off. Basically, an image reference is something that, currently in docker, can be `image`, `image:tag`, `image@sha256:…`, …

albers commented:
But doesn't that mean that in `<image-reference>|<image-id>`, `image-id` is already included in `image-reference`? Maybe `<image-reference>|<image-name>` would be a better wording here. Or just `<image-reference>`.

albers commented:
@vdemeester bash completion LGTM.

albers commented:
This is how the `before` and `since` filters are documented in the `docker-ps` man page:

          - before=(<container-name>|<container-id>)
          - since=(<container-name>|<container-id>)

vdemeester commented:
@albers you're right :-) For now I'll use the same notation as the ancestor filter of ps then (to be consistent)

vdemeester committed:
Add before and since filter to images

Add support for two now filter on the `images` command : `before` and
`since`. They work the same as the one on the `ps` command but for
images.

        $ docker images --filter before=myimage
        # display all images older than myimage
        $ docker images --filter since=myimage
        # display all images younger than myimage

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

sdurrheimer commented:
ZSH completion LGTM

SvenDowideit commented:
docs LGTM

thaJeztah commented:
LGTM 👍
